Managed ACL: Add AccessRestrictionList support

tleacmcsa opened the pull request:
Add Access Restriction support

- AccessRestriction class and unit tests
- Updates to access-control-server:
  - Added attributes:
    - CommissioningARL
    - ARL
  - Added events:
    - AccessRestrictionEntryChanged
    - FabricRestrictionReviewUpdate
  - Added commands:
    - ReviewFabricRestrictions
    - ReviewFabricRestrictionsResponse
  - Populate ARL from CommissioningARL upon commissioning complete event
- Added Server initialization parameters for:
  - AccessRestrictionProvider implementation
- Updated platform/linux examples:
  - ExampleAccessRestriction
  - AppMain optional support for AccessRestrictions
  - Options parses optional CommissioningARL entries
- Added include exception for vector and map (feature currently for higher resourced NIMs only)

tleacmcsa committed:
Add AccessRestrictionList support

chrisdecenzo commented on src/lib/support/DefaultStorageKeyAllocator.h:
```diff
@@ -129,6 +129,15 @@ class DefaultStorageKeyAllocator
     }
 
     static StorageKeyName AccessControlExtensionEntry(FabricIndex fabric) { return StorageKeyName::Formatted("f/%x/ac/1", fabric); }
+    static StorageKeyName AccessControlCommissioningArlEntry(size_t index)
```
can this be wrapped in the ifdef?

if CHIP_CONFIG_USE_ACCESS_RESTRICTIONS

tleacmcsa replied:
done

tleacmcsa replied:
i hit a build issue when including the header for the build flag, so I left it in for now.

chrisdecenzo commented on src/app/server/Server.h:
```diff
@@ -678,6 +689,8 @@ class Server
 
     Access::AccessControl mAccessControl;
     app::AclStorage * mAclStorage;
+    Access::AccessRestriction * mAccessRestriction;
```
wrap this in #if CHIP_CONFIG_USE_ACCESS_RESTRICTIONS?

tleacmcsa replied:
done

chrisdecenzo commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -64,8 +83,16 @@ class AccessControlAttribute : public AttributeAccessInterface, public EntryList
 public:
-    void OnEntryChanged(const SubjectDescriptor * subjectDescriptor, FabricIndex fabric, size_t index, const Entry * entry,
-                        ChangeType changeType) override;
+    void OnEntryChanged(const SubjectDescriptor * subjectDescriptor, FabricIndex fabric, size_t index,
```
Is there a really compelling reason to change the signature? Each refactor increases risks given the late landing of the feature.

tleacmcsa replied:
The signature didnt change, I just had to add namespaces to existing types to prevent conflicts.

chrisdecenzo commented on src/app/CommandHandlerImpl.cpp:
```diff
@@ -408,6 +408,7 @@ Status CommandHandlerImpl::ProcessCommandDataIB(CommandDataIB::Parser & aCommand
                                          .entityId    = concretePath.mCommandId };
         Access::Privilege requestPrivilege = RequiredPrivilege::ForInvokeCommand(concretePath);
         err                                = Access::GetAccessControl().Check(subjectDescriptor, requestPath, requestPrivilege);
+
```
can revert this change

tleacmcsa replied:
done

chrisdecenzo reviewed (commented):
first pass feedback

chrisdecenzo commented on examples/platform/linux/Options.cpp:
```diff
@@ -280,6 +292,9 @@ const char * sDeviceOptionHelp =
     "  --trace_decode <1/0>\n"
     "       A value of 1 enables traces decoding, 0 disables this (default 0).\n"
 #endif // CHIP_CONFIG_TRANSPORT_TRACE_ENABLED
+    "  --enable-access-restrictions <CommissioningARL JSON>\n"
+    "       Enable ACL cluster access restrictions with the provided JSON CommissioningARL. Example:\n"
+    "       \"[{\\\"endpoint\\\": 1,\\\"cluster\\\": 2,\\\"restrictions\\\": [{\\\"type\\\": 0,\\\"id\\\": 3}]}]\"\n"
```
```suggestion
#if CHIP_CONFIG_USE_ACCESS_RESTRICTIONS
    "  --enable-access-restrictions <CommissioningARL JSON>\n"
    "       Enable ACL cluster access restrictions with the provided JSON CommissioningARL. Example:\n"
    "       \"[{\\\"endpoint\\\": 1,\\\"cluster\\\": 2,\\\"restrictions\\\": [{\\\"type\\\": 0,\\\"id\\\": 3}]}]\"\n"
#endif // CHIP_CONFIG_USE_ACCESS_RESTRICTIONS
```

tleacmcsa replied:
Done

chrisdecenzo commented on src/access/AccessControl.h:
```diff
@@ -18,6 +18,12 @@
 
 #pragma once
 
+#include <access/AccessConfig.h>
```
should this be wrapped in the #define too?

tleacmcsa replied:
No, since that's where the #define is #defined :-)

chrisdecenzo commented on src/access/AccessRestriction.cpp:
```diff
@@ -0,0 +1,270 @@
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    // wildcard event subscriptions are allowed. All other requests require an entity id
```
```suggestion
    // wildcard event subscriptions are allowed since wildcard is only used when setting up the subscription and we want that request to succeed (when generating the report, this method will be called with the specific event id). All other requests require an entity id
```

Or some comment explaining why the wildcard check here should succeed

tleacmcsa replied:
done

chrisdecenzo commented on src/access/AccessRestriction.cpp:
```diff
@@ -0,0 +1,270 @@
+                if (requestPath.requestType == RequestType::kAttributeReadRequest ||
+                    requestPath.requestType == RequestType::kAttributeWriteRequest)
+                {
+                    return CHIP_ERROR_ACCESS_DENIED;
```
It would be helpful to ensure a log message gets generated in each of these "return CHIP_ERROR_ACCESS_DENIED" paths so its super clear in the device log that the denial was due to ARL

tleacmcsa replied:
I kept that logging in AcccessControl.  If the ARL check fails it will say it was restricted and details are printed above the call to ARL's check.

tcarmelveilleux replied:
You can have a CHIP_ERROR_ACCESS_DENIED_BY_ARL that is converted to a different log in AccessControl, and then further updated back down to CHIP_ERROR_ACCESS_DENIED. This would allow logging to be clearer on the reason, at least.

chrisdecenzo commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+         * @param [in] entry        The latest value of the entry which changed.
+         * @param [in] changeType   The type of change that occurred.
+         */
+        virtual void OnEntryChanged(FabricIndex fabricIndex, size_t index, SharedPtr<Entry> entry, ChangeType changeType) = 0;
```
I assume this leads to the cluster event getting triggered?

tleacmcsa replied:
Yes.  access-control-server is a listener and generates the event based on this callback.

tcarmelveilleux replied:
This is missing `MatterReportingAttributeChangeCallback` in the impl of the storage.

tleacmcsa replied:
Added to access-control-server.cpp.

chrisdecenzo commented on examples/platform/linux/ExampleAccessRestriction.h:
```diff
@@ -0,0 +1,66 @@
+public:
+    class ExampleAccessRestrictionEntryListener : public AccessRestriction::EntryListener
+    {
+        void OnEntryChanged(FabricIndex fabricIndex, size_t index, SharedPtr<AccessRestriction::Entry> entry,
```
what logic can you envision going into the OnEntryChanged event?

tleacmcsa replied:
None, I removed that example listener.

cecille commented on examples/network-manager-app/linux/args.gni:
```diff
@@ -22,3 +22,4 @@ chip_project_config_include_dirs = [
 ]
 
 chip_config_network_layer_ble = false
+chip_enable_access_restrictions = true
```
Putting a comment on a random file so it doesn't get lost. This is a really big change (25 files) to a cluster that is key to our security model. We need to give reviewers a chance to look this over correctly and we're right in the middle of a test event. Please wait on submitting this PR until folks have a reasonable chance to review (ie after the end of TE2, with a couple of days to allow sufficient reviews)

cecille commented on src/access/AccessControl.cpp:
```diff
@@ -352,6 +356,19 @@ CHIP_ERROR AccessControl::Check(const SubjectDescriptor & subjectDescriptor, con
+    if (mAccessRestriction != nullptr)
+    {
+        CHIP_ERROR result = mAccessRestriction->Check(subjectDescriptor, requestPath);
+        if (result != CHIP_NO_ERROR)
```
This is dangerous - if the access restriction delegate fails for any reason, access control fails out and denies access. This needs to be a specific check because this is a swappable component that may return ANY error code.

tleacmcsa replied:
What would you suggest?  This is part of the overall access control check which also blocks access upon any failure (other than CHIP_ERROR_NOT_IMPLEMENTED).  Since ARL is an extension of this processing, I don't see why it should behave differently.  If ARL is activated at compile time and populated with data at runtime, I believe it should block access upon any failure (except CHIP_ERROR_NOT_IMPLEMENTED which is indeed a gap right now).

cecille replied:
Because this is a component that is installed via SetAccessRestriction at runtime, you don't know what's on the other side of this. It could be failing with any error, you don't really know. If the delegate fails in a generic way, all clients are shut down from access to anything on the device including access over PASE.

tcarmelveilleux replied:
Denying access is better than allowing access to things that are not allowed. Definitely though the code must be well tested to avoid this fallback to always being in error.

cecille replied:
I agree that going conservative is better in general, but I'm worried that if there is a generic error it shuts down access to everything on the device, including clusters that are disallowed from being restricted by the ARL.

tleacmcsa replied:
That is the intent of this addition though.  It is a hard check and must pass if the feature is enabled.

cecille commented on src/access/AccessRestriction.cpp:
```diff
@@ -0,0 +1,280 @@
+
+CHIP_ERROR AccessRestriction::CreateFabricEntries(const FabricIndex fabricIndex)
+{
+    for (auto & entry : sCommissioningEntries)
```
why do all the fabrics match the commissioningARL?

This feels like example code that got mixed into the base implementation.

tleacmcsa replied:
When commissioning completes, the CommissioningARL entries are installed in the ARL attribute for the commissioning fabric.

cecille replied:
But this is not a requirement in the spec and it seems like the restrictions for commissioning would be more strict that the fabric restrictions. This is not an example implementation as far as I can tell - this is the handler for all devices. So having this here doesn't make sense.

tcarmelveilleux replied:
+1 please do not copy CommissioningARL to actual ARL

tleacmcsa replied:
Fixed/removed.  I misunderstood part of the spec.

cecille commented on src/access/AccessRestriction.cpp:
```diff
@@ -0,0 +1,280 @@
+
+void AccessRestriction::AddListener(EntryListener & listener)
+{
+    if (mListeners == nullptr)
```
I realize this was probably copied from the ACL code, but we have proper linked list support in lib/support now.

tleacmcsa replied:
Will fix.  Please see https://github.com/project-chip/connectedhomeip/issues/35165

cecille commented on src/access/AccessRestriction.cpp:
```diff
@@ -0,0 +1,280 @@
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    auto localEntry = MakeShared<Entry>(entry);
```
What are all the entities that are sharing this?

cecille commented on src/access/AccessRestriction.cpp:
```diff
@@ -0,0 +1,280 @@
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AccessRestriction::DeleteEntry(size_t index, const FabricIndex fabricIndex)
```
Is the idea here that you can delete entries base on the returned index value from the add? It seems like that would fall apart as soon as anything was deleted.

Can this function signature be changed to not use the index (and above not to return it)? Index is not a stable identifier in a vector.

tleacmcsa replied:
The data management model has been greatly simplified and this is now gone.

cecille commented on src/access/AccessRestriction.cpp:
```diff
@@ -0,0 +1,280 @@
+        }
+    }
+
+    for (auto & entry : mFabricEntries[subjectDescriptor.fabricIndex])
```
This does not work for commissioning - you will need to check specifically for the pase fabric index, but then ALSO check whether there is an ongoing commissioning on this fabric index in order to defer to the CommissioningARL entry.

tleacmcsa replied:
This should be fixed now.

cecille commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+{
+public:
+    static constexpr size_t kNumberOfFabrics      = CHIP_CONFIG_MAX_FABRICS;
+    static constexpr size_t kEntriesPerFabric     = CHIP_CONFIG_ACCESS_RESTRICTION_MAX_ENTRIES_PER_FABRIC;
```
Is this enforced by the code anywhere? As far as I can tell, you can just keep pushing new entries onto the vector until your memory runs out.

tleacmcsa replied:
Will track in https://github.com/project-chip/connectedhomeip/issues/35165

cecille commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+public:
+    static constexpr size_t kNumberOfFabrics      = CHIP_CONFIG_MAX_FABRICS;
+    static constexpr size_t kEntriesPerFabric     = CHIP_CONFIG_ACCESS_RESTRICTION_MAX_ENTRIES_PER_FABRIC;
+    static constexpr size_t kRestrictionsPerEntry = CHIP_CONFIG_ACCESS_RESTRICTION_MAX_RESTRICTIONS_PER_ENTRY;
```
same q here

tleacmcsa replied:
Will track in https://github.com/project-chip/connectedhomeip/issues/35165

cecille commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+    static constexpr size_t kNumberOfFabrics      = CHIP_CONFIG_MAX_FABRICS;
+    static constexpr size_t kEntriesPerFabric     = CHIP_CONFIG_ACCESS_RESTRICTION_MAX_ENTRIES_PER_FABRIC;
+    static constexpr size_t kRestrictionsPerEntry = CHIP_CONFIG_ACCESS_RESTRICTION_MAX_RESTRICTIONS_PER_ENTRY;
+
```
Note that for above you will either need to update our resources certification test to check this (TC-RR-1.1) or handle in a separate test case. RR-1.1 handles the ACL checks.

cecille commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+     * @param [out] begin iterator pointing to the beginning of the entries
+     * @param [out] end iterator pointing to the end of the entries
+     */
+    CHIP_ERROR Entries(const FabricIndex fabricIndex, EntryIterator & begin, EntryIterator & end)
```
Does this properly handle delete calls while holding the iterator?

tleacmcsa replied:
This data management model has been greatly simplified and this isnt an issue anymore.

cecille commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+     * @param [in]  entry       The entry to add to the commissioning access restriction list.
+     * @return                  CHIP_NO_ERROR if the entry was successfully created, or an error code.
+     */
+    static CHIP_ERROR CreateCommissioningEntry(SharedPtr<Entry> entry);
```
why is the API different for CreateCommissioningEntry and CreateEntry (aside from the fabricIndex)

tleacmcsa replied:
This data management model has been greatly simplified and this isnt an issue anymore.

cecille reviewed (commented):
(self note - made it down to the tests, didn't review tests and onwards yet)

tleacmcsa commented on examples/platform/linux/Options.cpp:
```diff
@@ -280,6 +292,11 @@ const char * sDeviceOptionHelp =
+#if CHIP_CONFIG_USE_ACCESS_RESTRICTIONS
+    "  --enable-access-restrictions <CommissioningARL JSON>\n"
+    "       Enable ACL cluster access restrictions with the provided JSON CommissioningARL. Example:\n"
+    "       \"[{\\\"endpoint\\\": 1,\\\"cluster\\\": 2,\\\"restrictions\\\": [{\\\"type\\\": 0,\\\"id\\\": 3}]}]\"\n"
```
this option enables the overall ARL feature which works by populating the CommissioningARL attribute.  Once the device is commissioned the CommissioningARL attribute is copied to the commissioning fabric's ARL attribute.  There is no other way provided in the example apps with this PR to directly manipulate the ARL besides this and the behavior of the ReviewFabricRestrictions command which simply removes all ARL entries for the calling fabric.

I can't think of a better name, but I'm certainly open to suggestions.

cecille replied:
Can you just add that as documentation maybe?

That being said, you probably DO want a way to set these separately to ensure that the restrictions are being applied appropriately. Right now, i don't believe they are.

tleacmcsa replied:
I fixed this up.  Now we have --commissioning-arl-entries and --arl-entries

tleacmcsa commented on scripts/tools/check_includes_config.py:
```diff
@@ -185,4 +185,5 @@
     'src/app/icd/client/DefaultICDStorageKey.h': {'vector'},
     'src/controller/CHIPDeviceController.cpp': {'string'},
     'src/qrcodetool/setup_payload_commands.cpp': {'string'},
+    'src/access/AccessRestriction.h': {'vector', 'map'},
```
Correct.  I figured with the push to restrict to NIM only in 1.4, this shouldn't be an issue due to the typical resources available to these devices.

tleacmcsa replied:
Can we resolve this comment?

tleacmcsa committed:
Update src/access/AccessConfig.h

Co-authored-by: C Freeman <cecille@google.com>

tcarmelveilleux commented on examples/network-manager-app/linux/args.gni:
```diff
@@ -22,3 +22,4 @@ chip_project_config_include_dirs = [
 ]
 
 chip_config_network_layer_ble = false
+chip_enable_access_restrictions = true
```
Please indicate with a comment that this is what enables the ARL support and the NIM sample app requires it.

tleacmcsa replied:
done

tcarmelveilleux commented on src/app/server/ArlStorage.h:
```diff
@@ -0,0 +1,148 @@
+     * Typically used temporarily on the stack to encode:
+     * - source: system level access restriction entry
+     * - staging: generated cluster level code
+     * - destination: TLV
```
There is no `destination` anywhere.

tleacmcsa replied:
fixed

tcarmelveilleux commented on src/app/server/ArlStorage.h:
```diff
@@ -0,0 +1,148 @@
+        using StagingRestriction = Clusters::AccessControl::Structs::AccessRestrictionStruct::Type;
+
+    public:
+        EncodableEntry(std::shared_ptr<Entry> entry) : mEntry(entry) {}
```
Why does this need to be a shared_ptr?

tleacmcsa replied:
I have removed all shared_ptrs

tcarmelveilleux commented on src/app/server/ArlStorage.h:
```diff
@@ -0,0 +1,148 @@
+        EncodableEntry(std::shared_ptr<Entry> entry) : mEntry(entry) {}
+
+        /**
+         * Constructor-provided entry is staged into a staging entry,
+         * which is then encoded into a writer.
+         */
```
Please document the arguments for these methods, rather than the implementation. The staging is an implementation detail and not described.

tleacmcsa replied:
This should be good now (ArlEncoder.h)

tcarmelveilleux commented on src/app/server/ArlStorage.h:
```diff
@@ -0,0 +1,148 @@
+     * loaded from persistent storage. A half-open range of fabrics [first, last) is provided
+     * so this can be done on a per-fabric basis.
+     *
+     * Implementations should also install an entry change listener on AccessControl to maintain
+     * ARL entries in persistent storage as they are changed.
```
How is this done or supposed to be done.

I would have expected `ArlStorage` interface to have accessors for the ARL entries.

tleacmcsa replied:
This should be more clear now.  AccessRestrictionProvider is the single source-of-truth for the data.

tcarmelveilleux commented on src/app/server/DefaultArlStorage.h:
```diff
@@ -0,0 +1,37 @@
+namespace chip {
+namespace app {
+
+class DefaultArlStorage : public ArlStorage
```
Please call this `ExampleArlStorage`. Calling it default assumes many implementers would use it as-is, but ARL handling is 100% product specific due to how overrides can be requested, so this should be an example only.

tcarmelveilleux commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+namespace Access {
+
+template <typename T>
+using SharedPtr = chip::Platform::SharedPtr<T>;
```
If a shared_ptr is going to use, let's use std.

See:

```
template <typename T>
using SharedPtr = std::shared_ptr<T>;

template <typename T, typename... Args>
inline SharedPtr<T> MakeShared(Args &&... args)
{
    return SharedPtr<T>(New<T>(std::forward<Args>(args)...), Deleter<T>());
}
```

This legacy of pretending that platform could override impl of shared pointer is not a thing.

tleacmcsa replied:
I have removed all shared_ptrs

tcarmelveilleux commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+         * @param [in] instruction  The instructions to be displayed to the user.
+         * @param [in] redirectUrl  An optional URL to redirect the user to for more information.  May be null.
+         */
+        virtual void OnFabricRestrictionReviewUpdate(FabricIndex fabricIndex, uint64_t token, const char * instruction,
```
`const char*` means that the backing be a null terminated string. Suggest CharSpan, which supports safe length handling and backing by different storage.

tleacmcsa replied:
done.

tcarmelveilleux commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+private:
+    SharedPtr<Entry> GetEntry(FabricIndex fabricIndex, size_t index);
+
+    static std::vector<Platform::SharedPtr<Entry>> sCommissioningEntries;
```
This should not be static. This should be a member just like mFabricEntries

tleacmcsa replied:
done.

tcarmelveilleux commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+    void ClearData();
+
+private:
+    SharedPtr<Entry> GetEntry(FabricIndex fabricIndex, size_t index);
```
This API would be easier to back if GetEntry simply returned a new entry. The copy costs are likely very small compared to the number of calls.

tleacmcsa replied:
data management apis have been redone.

tcarmelveilleux commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+     * @param [in] arl          An optinal list of access restriction entries to review.  If null, all entries will be reviewed.
+     * @return                  CHIP_NO_ERROR if the review was successfully requested, or an error code if the request failed.
+     */
+    virtual CHIP_ERROR DoRequestFabricRestrictionReview(const FabricIndex fabricIndex, uint64_t token,
```
Instead of using inheritance here, it would be best if this was delegated to the ARL storage or equivalnt delegate. In fact, calling ArlStorage a storage class makes it seems like it needs to be always backed in memory and kept in sync, when it would be possible to fully delegate the access to something that never maintains the issues fully in memory and implements all rules via further delegation.

tleacmcsa replied:
I believe now that the storage classes have been removed, much of this comment is resolved.  It still uses inheritance for now.

tleacmcsa replied:
I believe now that the storage classes have been removed, much of this comment is resolved.  It still uses inheritance for now.

tcarmelveilleux commented on src/access/AccessControl.h:
```diff
@@ -627,6 +633,13 @@ class AccessControl
 
+#if CHIP_CONFIG_USE_ACCESS_RESTRICTIONS
+    // Set an optional AcceessRestriction object for MNGD feature.
+    void SetAccessRestriction(AccessRestriction * accessRestriction) { mAccessRestriction = accessRestriction; }
```
AccessRestriction --> AccessRestrictionProvider

tleacmcsa replied:
done

tcarmelveilleux commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -473,6 +601,16 @@ CHIP_ERROR AccessControlAttribute::Write(const ConcreteDataAttributePath & aPath
+#if CHIP_CONFIG_USE_ACCESS_RESTRICTIONS
+void OnPlatformEventHandler(const DeviceLayer::ChipDeviceEvent * event, intptr_t arg)
+{
+    if (event->Type == DeviceLayer::DeviceEventType::kCommissioningComplete)
```
This happens on CommissioningComplete command, which ends a fail-safe, which can occur outside of "commissioning flows" (as in joing a new fabric) and can arise when reconfiguring networking or doing an UpdateNOC, both of which require ArmFailSafe/CommissioningComplete pairs. 

I don't think there is any need to be stateful on "commissioning flow".

cecille replied:
The spec actually does say that review fabric restrictions is disallowed during commissioning. If the device adds ARL attributes before the fabric is added then it has to be added to the failsafe context in order to be properly rolled back. When I last talked to Chris, he said he was opting to disallow ARL changes during commissioning rather than play with the failsafe context and therefore the CommissioningARL restrictions would continue to be in effect until the commissioning complete was received IF the accessing fabric matches the fabric in the failsafe context.

The switch over from CommissioningARL to ARL I think IS required to happen at commissioning complete if the client is on the new fabric. Because otherwise there's a window of time where restrictions don't apply (ie, you're on a new fabric, but new fabric ARL entires have not yet been added). If you're swapping over to ARL directly after AddNOC, then the ARL entries need to be part of the failsafe. 

Incidentally, the description of CommissioningARL attribute in the text contradicts the description of the CommissioningARL in the ARL attribute. So that needs clarification as to exactly when this switch over applies.

cecille replied:
Which brings up another question - are the ARL entries for a fabric properly cleared when a fabric is removed, and are there tests for that?

tcarmelveilleux replied:
The ARL entries are not "stored" persistently (or shouldn't be, in real products). They should reflect a current state.

Regarding the race window, I think it's fair that the CommissioningARL applies:

- Whenever the session is a PASE session without an accessing fabric
- Whenever the session is a CASE session for a **pending** fabric not yet committed by CommissioningComplete (e.g. after an AddNOC, in the context of the new fabric index)

Then, basically, the CommissioningARL ought to block all aspects of the Node which are not accessible unless there would be an override. This can be known a-priori.

tleacmcsa replied:
This should be resolved now.

tcarmelveilleux commented on src/access/AccessRestriction.cpp:
```diff
@@ -0,0 +1,280 @@
+void AccessRestriction::ClearData()
+{
+    // iterate over mFabricEntries and call DeleteEntry for each entry
+    for (auto & fabricEntry : mFabricEntries)
+    {
+        for (size_t i = 0; i < fabricEntry.second.size(); i++)
+        {
+            DeleteEntry(i, fabricEntry.first);
+        }
+    }
```
This needs to be done in reverse order otherwise you will immediately run through the end.

tleacmcsa replied:
This is no longer relevant given the data model management change.

tcarmelveilleux commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+     * @param [in]  fabricIndex The index of the fabric in which the entry should be added.
+     * @return                  CHIP_NO_ERROR if the entry was successfully created, or an error code.
+     */
+    CHIP_ERROR CreateEntry(size_t * index, const Entry & entry, const FabricIndex fabricIndex);
```
Suggest an overall `SetEntries`/`SetCommissioningEntries`.

This has the benefits of:

- Removals can be computed by set difference (if needed, would argue for removal of deletion events)
- Additions/updates the same
- You now know when the  list is done updating so you can call `MatterReportingAttributeChangeCallback`
- You can easily schedule a single setter lambda with a unique_ptr arg of the whole dataset to make sure it's called in correct context.

If you do this, you do have to normalize the list:

- The shadow copy/internal copy must always be kept in sorted order so that it's possible to tell the difference between changed and unchanged of doing SetXxx.
  - A unit test should be done that has out of order same entries and not report it.
  - That avoids over-reporting same data  (same semantic meaning) but different order because externally generated via things like unordered maps in the runtime)

tleacmcsa replied:
I believe @chrisdecenzo is going to remove the related entry event (updated/deleted).  I did implement the two new Set methods to change both lists completely.  I believe if the simplification is made to the spec, this should be resolved.

cecille commented on src/access/AccessRestriction.cpp:
```diff
@@ -0,0 +1,280 @@
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+    }
+
+    for (auto & entry : mFabricEntries[subjectDescriptor.fabricIndex])
```
With the spec the way it is, CommissioningARL check is slightly more complicated because you need to check both
- no fabric index
- if failsafeContext is armed and accessing fabric index == failsafe fabric index and AddNoc has been invoked.

tleacmcsa committed:
Reworked data manipulators and other cleanup

bzbarsky-apple commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,193 @@
+        listener->RestrictionListChanged(fabricIndex);
+    }
+
+    return CHIP_NO_ERROR;
```
But ... don't we need to actually store `entries` somehere?

How does this pass tests?  This suggests the tests are quite insufficient.

tleacmcsa replied:
Fixed and agree on the tests.

bzbarsky-apple commented on src/access/AccessRestrictionProvider.h:
```diff
@@ -0,0 +1,225 @@
+     * @param [in] fabricIndex  The index of the fabric for which to create entries.
+     * @param [in] entries  The entries to set for the fabric.
+     */
+    CHIP_ERROR SetEntries(const FabricIndex, const std::vector<Entry> & entries);
```
I don't see where entries are cleared on fabric removal.  That needs to happen, no?

tleacmcsa replied:
That is up to the app, which is responsible for managing the entries.

bzbarsky-apple replied:
We know the entries have to be removed when the fabric is removed.  Why are we forcing the app to remember to do that?

tleacmcsa replied:
The app is completely responsible for managing the ARL data.  It is not persisted by the SDK.  If we make the SDK also responsible for manipulating the data, there will be contention.

bzbarsky-apple commented on src/access/AccessRestrictionProvider.h:
```diff
@@ -0,0 +1,225 @@
+     * Get the restriction entries for a fabric.
+     *
+     * @param [in]  fabricIndex the index of the fabric for which to get entries.
+     * @param [out] entries reference to a const vector to hold the entries.
```
Needs to clearly say how long the pointer is valid for.

tleacmcsa replied:
Fixed

bzbarsky-apple commented on src/access/AccessRestrictionProvider.h:
```diff
@@ -0,0 +1,225 @@
+     * @param [in] arl          An optinal list of access restriction entries to review.  If null, all entries will be reviewed.
+     * @return                  CHIP_NO_ERROR if the review was successfully requested, or an error code if the request failed.
+     */
+    virtual CHIP_ERROR DoRequestFabricRestrictionReview(const FabricIndex fabricIndex, uint64_t token,
```
So the way this is currently implemented, the client can get the event before it gets the command response.  This is fine, and unavoidable, but the spec should probably have some informative language making this clear to spec readers.

tleacmcsa replied:
Resolving and tracking with https://github.com/project-chip/connectedhomeip/issues/35165

bzbarsky-apple commented on src/access/AccessRestrictionProvider.h:
```diff
@@ -0,0 +1,225 @@
+
+    uint64_t mNextToken   = 1;
+    Listener * mListeners = nullptr;
+    std::vector<Entry> mCommissioningEntries;
+    std::map<FabricIndex, std::vector<Entry>> mFabricEntries;
```
So these will lead to crashes on out-of-heap, right?

How will this work for constrained devices that might want to ensure static allocation for such things?

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,100 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+
+        for (auto & entry : entries)
+        {
+            ArlStorage::EncodableEntry encodableEntry(entry);
+            ReturnErrorOnFailure(encoder.Encode(encodableEntry));
```
This looks wrong.  `ArlStorage::EncodableEntry` encodes a fabric-scoped struct, but CommissioningARL consists of non-fabric-scoped structs.

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,100 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+
+    MatterReportingAttributeChangeCallback(0, AccessControlCluster::Id, AccessControlCluster::Attributes::Arl::Id);
+
+    ArlChangedEvent event{ .fabricIndex = fabricIndex };
```
So spec says these events are generated "whenever its ARL attribute data is changed by the device."

It doesn't define the fabric-association of the events, so this code is just making up some non-spec-defined behavior.  Maybe that behavior is the right one; I can't tell.  Spec needs to define this.

tleacmcsa replied:
This event has now been removed from the spec so I will remove it from the code shortly.

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,100 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+
+    if (instruction != nullptr)
+    {
+        event.instruction.SetNonNull(chip::CharSpan(instruction, strnlen(instruction, kMaxInstructionStringLength)));
```
Why all the extra "chip::" prefixes in various places?

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -481,4 +614,68 @@ void MatterAccessControlPluginServerInitCallback()
+            }
+            entry.restrictions.push_back(restriction);
+        }
+
```
Where is the check for malformed TLV (i.e. restrictionIter.GetStatus())?

tleacmcsa replied:
This should be resolved now (for the DecodableList, not the std::vector)

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -481,4 +614,68 @@ void MatterAccessControlPluginServerInitCallback()
+
+        entries.push_back(entry);
+    }
+
```
Again, where is the "iteration just failed" check?

tleacmcsa replied:
This should be resolved now (for the DecodableList, not the std::vector)

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -481,4 +614,68 @@ void MatterAccessControlPluginServerInitCallback()
+    {
+        ChipLogError(DataManagement, "AccessControlCluster: restriction check failed: %" CHIP_ERROR_FORMAT, err.Format());
+
+        // return error to client?
```
Yes, yes you have to.

tleacmcsa replied:
Done

bzbarsky-apple commented on src/app/server/ArlStorage.cpp:
```diff
@@ -0,0 +1,158 @@
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Convert(AccessRestrictionProvider::Type from, StagingRestrictionType & to)
```
If we have this function, why are we just directly static_casting between the two in `emberAfAccessControlClusterReviewFabricRestrictionsCallback`?

bzbarsky-apple commented on src/app/server/ArlStorage.cpp:
```diff
@@ -0,0 +1,158 @@
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ArlStorage::EncodableEntry::EncodeForWrite(TLV::TLVWriter & writer, TLV::Tag tag) const
```
So this is used to store the entries in persistent storage, right?

This part is very much not making sense to me.  `mStagingEntry.EncodeForWrite` will not encode a fabric index (it's designed to be used for attribute writes, and per spec those do not include fabric indices on fabric scoped structs).  Then when we read from storage we do a `Decode()`, never set the fabric index on the struct (right?) but then later code assumes those structs have a fabric index set.

This all needs a _lot_ more tests.

tleacmcsa replied:
Fixed.

bzbarsky-apple commented on src/app/server/DefaultArlStorage.cpp:
```diff
@@ -0,0 +1,229 @@
+
+DATA                                    C   T   L   V   NOTES
+structure (anonymous)                   1               0x15
+  field 0 endpoint                      1   1       1
```
The V here is wrong.  Should be 2.

tleacmcsa replied:
This code has been removed.

bzbarsky-apple commented on src/lib/core/CHIPConfig.h:
```diff
@@ -1206,6 +1206,27 @@ extern const char CHIP_NON_PRODUCTION_MARKER[];
+ * fabric in the access control code's ARL attribute.
+ */
+#ifndef CHIP_CONFIG_ACCESS_RESTRICTION_MAX_ENTRIES_PER_FABRIC
+#define CHIP_CONFIG_ACCESS_RESTRICTION_MAX_ENTRIES_PER_FABRIC 10
```
I don't see floors for these constants defined in the spec.  Or ceilings.  There should probably be such definitions.

tleacmcsa replied:
Tracking with https://github.com/project-chip/connectedhomeip/issues/35165

tleacmcsa committed:
Fixed encode/decode so reading CommissioningARL and Arl attributes work

tleacmcsa committed:
Merge branch 'project-chip:master' into arl-feature

tleacmcsa commented on src/access/AccessRestriction.h:
```diff
@@ -0,0 +1,288 @@
+     * @return                  CHIP_NO_ERROR if the review was successfully requested, or an error code if the request failed.
+     */
+    virtual CHIP_ERROR DoRequestFabricRestrictionReview(const FabricIndex fabricIndex, uint64_t token,
+                                                        const std::vector<Entry> & arl) = 0;
```
This is the ARL argument from the ReviewFabricRestrictions command which is an optional list of ARL entries to review.

tleacmcsa commented on examples/platform/linux/AppMain.cpp:
```diff
@@ -602,12 +602,12 @@ void ChipLinuxAppMainLoop(AppMainLoopImplementation * impl)
-        {
-            VerifyOrDie(AccessRestriction::CreateCommissioningEntry(entry) == CHIP_NO_ERROR);
-        }
+        auto exampleAccessRestrictionProvider = new ExampleAccessRestrictionProvider();
```
done.

tleacmcsa committed:
Reworked ARL storage

Previously ARL related data was persisted in KVS.  This has been
removed and now the responsibility for managing/maintaining the
related data (CommissioningARL and ARL attributes) is up to the
app to set on AccessRestrictionProvider class.

tleacmcsa committed:
Review fixes

cleanup ArlEncoder interface.
return error to client if arl review request fails
return token to client in FabricRestrictionReviewUpdate

tleacmcsa commented on examples/platform/linux/ExampleAccessRestrictionProvider.h:
```diff
@@ -0,0 +1,54 @@
+        // this example simply removes all restrictions and will generate AccessRestrictionEntryChanged events
+        Access::GetAccessControl().GetAccessRestrictionProvider()->SetEntries(fabricIndex, std::vector<Entry>{});
+
+        chip::app::Clusters::AccessControl::Events::FabricRestrictionReviewUpdate::Type event{ .fabricIndex = fabricIndex };
```
Fixed.

tleacmcsa committed:
Fixed GetEntries vector pointer arg

tleacmcsa committed:
Updated core restriction logic/integration

restyled-commits committed:
Restyled by clang-format

tleacmcsa committed:
fixed include check for renamed AccessRestrictionProvider.h file

tleacmcsa commented on src/access/tests/TestAccessRestriction.cpp:
```diff
@@ -0,0 +1,463 @@
+AccessControl accessControl;
+TestAccessRestrictionImpl accessRestriction;
+
+constexpr ClusterId kNetworkCommissioningCluster = 0x0000'0031; // must not be blocked by access restrictions on any endpoint
```
Resolving and tracking with https://github.com/project-chip/connectedhomeip/issues/35165

tleacmcsa commented on src/access/tests/TestAccessRestriction.cpp:
```diff
@@ -0,0 +1,463 @@
+      .requestPath       = { .cluster = 1, .endpoint = 1, .requestType = RequestType::kEventReadOrSubscribeRequest, .entityId = 1 },
+      .privilege         = Privilege::kAdminister,
+      .allow             = true },
+};
```
Resolving and tracking with https://github.com/project-chip/connectedhomeip/issues/35165

tleacmcsa commented on src/access/tests/TestAccessRestriction.cpp:
```diff
@@ -0,0 +1,463 @@
+        .fabricIndex = 2,
+        .privilege   = Privilege::kAdminister,
+        .authMode    = AuthMode::kCase,
+        .subject     = kOperationalNodeId2,
```
Resolving and tracking with https://github.com/project-chip/connectedhomeip/issues/35165

tcarmelveilleux commented on examples/platform/linux/Options.cpp:
```diff
@@ -320,6 +342,39 @@ const char * sDeviceOptionHelp =
 #endif
     "\n";
 
+#if CHIP_CONFIG_USE_ACCESS_RESTRICTIONS
+bool ParseAccessRestrictionEntriesFromJson(const char * jsonString, std::vector<AccessRestrictionProvider::Entry> & entries)
+{
+    Json::Value root;
+    Json::Reader reader;
+    VerifyOrReturnValue(reader.parse(jsonString, root), false);
+
+    for (Json::Value::const_iterator eIt = root.begin(); eIt != root.end(); eIt++)
+    {
+        AccessRestrictionProvider::Entry entry;
+
+        entry.endpointNumber = static_cast<EndpointId>((*eIt)["endpoint"].asUInt());
+        entry.clusterId      = static_cast<ClusterId>((*eIt)["cluster"].asUInt());
+
+        Json::Value restrictions = (*eIt)["restrictions"];
+        for (Json::Value::const_iterator rIt = restrictions.begin(); rIt != restrictions.end(); rIt++)
+        {
+            AccessRestrictionProvider::Restriction restriction;
+            restriction.restrictionType = static_cast<AccessRestrictionProvider::Type>((*rIt)["type"].asInt());
+            if ((*rIt).isMember("id"))
+            {
+                restriction.id.SetValue((*rIt)["id"].asUInt());
+            }
+            entry.restrictions.push_back(restriction);
+        }
+
+        entries.push_back(entry);
+    }
+
+    return true;
+}
+#endif // CHIP_CONFIG_USE_ACCESS_RESTRICTIONS
```
Suggest making a separate file that contains this so it's easier to move it later to a more common spot. Can be done in a follow-up.

tleacmcsa replied:
See https://github.com/project-chip/connectedhomeip/issues/35189

tcarmelveilleux commented on examples/platform/linux/Options.cpp:
```diff
@@ -529,6 +584,27 @@ bool HandleOption(const char * aProgram, OptionSet * aOptions, int aIdentifier,
 #endif // CHIP_CONFIG_TRANSPORT_TRACE_ENABLED
 
+#if CHIP_CONFIG_USE_ACCESS_RESTRICTIONS
+    case kDeviceOption_CommissioningArlEntries: {
+        std::vector<AccessRestrictionProvider::Entry> entries;
+        retval = ParseAccessRestrictionEntriesFromJson(aValue, entries);
+        if (retval)
+        {
+            LinuxDeviceOptions::GetInstance().commissioningArlEntries.SetValue(std::move(entries));
+        }
+    }
+    break;
+    case kDeviceOption_ArlEntries: {
+        std::vector<AccessRestrictionProvider::Entry> entries;
+        retval = ParseAccessRestrictionEntriesFromJson(aValue, entries);
+        if (retval)
+        {
+            LinuxDeviceOptions::GetInstance().arlEntries.SetValue(std::move(entries));
+        }
+    }
```
This is extremely cumbersome to put on the command line... Suggest using file paths and loading from file.

tleacmcsa replied:
No kidding!  Thats a good enhancement idea.  I'll fix that up next week.

tcarmelveilleux replied:
Please file an issue and add `TODO(#xxxxx): change to using a path instead`

tleacmcsa replied:
Done

tleacmcsa committed:
M-ACL updates

- refactored AccessControl::Check into CheckACL and CheckARL
- added placeholders for the upcoming CHIP_ERROR_ACCESS_RESTRICTED_BY_ARL
- extracted ARL exception processing to standalone class for better
  testing

tcarmelveilleux committed:
Add plumbing for subject descriptor IsCommissioning field

- Make session manager update that state on a message-per-message basis
- Add tests

Missing test: MRP test against a not-yet-committed fabric over CASE showing
that IsCommissioning is true.

tleacmcsa committed:
Merge branch 'project-chip:master' into arl-feature

tcarmelveilleux committed:
Fix crash

tleacmcsa committed:
Use new IsCommissioning in ARL check

tleacmcsa commented on examples/platform/linux/Options.cpp:
```diff
@@ -320,6 +342,39 @@ const char * sDeviceOptionHelp =
+        for (Json::Value::const_iterator rIt = restrictions.begin(); rIt != restrictions.end(); rIt++)
+        {
+            AccessRestrictionProvider::Restriction restriction;
+            restriction.restrictionType = static_cast<AccessRestrictionProvider::Type>((*rIt)["type"].asInt());
```
Fixed

bzbarsky-apple commented on src/app/clusters/access-control-server/ArlEncoder.cpp:
```diff
@@ -0,0 +1,152 @@
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ArlEncoder::EncodableEntry::EncodeForWrite(TLV::TLVWriter & writer, TLV::Tag tag) const
```
I don't think this EncodeForWrite method is used anywhere.   Would suggest removing it for now.

tcarmelveilleux replied:
+1

tleacmcsa replied:
done

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,99 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+}
+void AccessControlAttribute::CommissioningRestrictionListChanged()
+{
+    MatterReportingAttributeChangeCallback(0, AccessControlCluster::Id, AccessControlCluster::Attributes::CommissioningARL::Id);
```
```suggestion
    MatterReportingAttributeChangeCallback(kRootEndpointId, AccessControlCluster::Id, AccessControlCluster::Attributes::CommissioningARL::Id);
```

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,99 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+{
+    CHIP_ERROR err;
+
+    MatterReportingAttributeChangeCallback(0, AccessControlCluster::Id, AccessControlCluster::Attributes::Arl::Id);
```
```suggestion
    MatterReportingAttributeChangeCallback(kRootEndpointId, AccessControlCluster::Id, AccessControlCluster::Attributes::Arl::Id);
```

tleacmcsa replied:
done

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,99 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+
+    if (instruction != nullptr)
+    {
+        event.instruction.SetNonNull(chip::CharSpan(instruction, strnlen(instruction, kMaxInstructionStringLength)));
```
```suggestion
        event.instruction.SetNonNull(CharSpan(instruction, strnlen(instruction, kMaxInstructionStringLength)));
```
This was in the previous set of review comemnts....

tleacmcsa replied:
done

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,99 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+
+    if (redirectUrl != nullptr)
+    {
+        event.redirectURL.SetNonNull(chip::CharSpan(redirectUrl, strnlen(redirectUrl, kMaxRedirectUrlStringLength)));
```
```suggestion
        event.redirectURL.SetNonNull(CharSpan(redirectUrl, strnlen(redirectUrl, kMaxRedirectUrlStringLength)));
```

tleacmcsa replied:
done

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,99 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+    }
+
+    EventNumber eventNumber;
+    SuccessOrExit(err = LogEvent(event, 0, eventNumber));
```
```suggestion
    SuccessOrExit(err = LogEvent(event, kRootEndpointId, eventNumber));
```

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,99 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+    ArlChangedEvent event{ .fabricIndex = fabricIndex };
+
+    EventNumber eventNumber;
+    SuccessOrExit(err = LogEvent(event, 0, eventNumber));
```
```suggestion
    SuccessOrExit(err = LogEvent(event, kRootEndpointId, eventNumber));
```

tleacmcsa replied:
done

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -481,4 +613,72 @@ void MatterAccessControlPluginServerInitCallback()
+
+        entries.push_back(entry);
+    }
+
```
This code is still missing the "did we get invalid input checks".  See https://github.com/project-chip/connectedhomeip/pull/34932/#discussion_r1725382119 and https://github.com/project-chip/connectedhomeip/pull/34932/#discussion_r1725382583 from my previous review.

tleacmcsa replied:
There are two iterators here, one for the DecodableList and one for a std:vector.  I added a check for error via GetStatus on the former, but I dont think there is anything to do for the latter.

You had mentioned some additional complexity around loading more entries.  I have not yet handled that complexity and request I fix in an immediate follow up so I have time to review and understand that mechanism.  It should work for "small" lists in the interim.

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -481,4 +613,72 @@ void MatterAccessControlPluginServerInitCallback()
+    else
+    {
+        ChipLogError(DataManagement, "AccessControlCluster: restriction review failed: %" CHIP_ERROR_FORMAT, err.Format());
+        commandObj->AddStatus(commandPath, Protocols::InteractionModel::Status::Failure);
```
```suggestion
        commandObj->AddStatus(commandPath, Protocols::InteractionModel::ClusterStatusCode(err));
```
in case it encodes something useful.

tleacmcsa replied:
done

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -481,4 +613,72 @@ void MatterAccessControlPluginServerInitCallback()
+    CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
+    const Clusters::AccessControl::Commands::ReviewFabricRestrictions::DecodableType & commandData)
+{
+    if (commandPath.mEndpointId != 0)
```
```suggestion
    if (commandPath.mEndpointId != kRootEndpointId)
```

tleacmcsa replied:
done

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -481,4 +613,72 @@ void MatterAccessControlPluginServerInitCallback()
+            AccessRestrictionProvider::Restriction restriction;
+            if (ArlEncoder::Convert(restrictionIter.GetValue().type, restriction.restrictionType) != CHIP_NO_ERROR)
+            {
+                ChipLogError(DataManagement, "AccessControlCluster: invalid restriction type conversion");
```
This is not spec-compliant. Per spec, an out-of-range value should respond with an error, no?

tcarmelveilleux replied:
Same: InvalidCommand

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -481,4 +613,72 @@ void MatterAccessControlPluginServerInitCallback()
+    if (commandPath.mEndpointId != 0)
+    {
+        ChipLogError(DataManagement, "AccessControlCluster: invalid endpoint in ReviewFabricRestrictions request");
+        return true;
```
This needs to respond with an error.

bzbarsky-apple commented on src/credentials/FabricTable.h:
```diff
@@ -736,6 +736,18 @@ class DLL_EXPORT FabricTable
     bool HasOperationalKeyForFabric(FabricIndex fabricIndex) const;
 
+    /**
+     * @brief If a newly-added fabric is pending, this returns its index, or kUndefinedFabricIndex in none are pending.
```
```suggestion
     * @brief If a newly-added fabric is pending, this returns its index, or kUndefinedFabricIndex if none are pending.
```

tleacmcsa replied:
done

bzbarsky-apple commented on src/lib/support/DefaultStorageKeyAllocator.h:
```diff
@@ -130,6 +130,16 @@ class DefaultStorageKeyAllocator
 
     static StorageKeyName AccessControlExtensionEntry(FabricIndex fabric) { return StorageKeyName::Formatted("f/%x/ac/1", fabric); }
 
+    static StorageKeyName AccessControlCommissioningArlEntry(size_t index)
```
Unused; please remove.

tleacmcsa replied:
done

bzbarsky-apple commented on src/lib/support/DefaultStorageKeyAllocator.h:
```diff
@@ -130,6 +130,16 @@ class DefaultStorageKeyAllocator
+        return StorageKeyName::Formatted("g/car/%x", static_cast<unsigned>(index));
+    }
+
+    static StorageKeyName AccessControlArlEntry(FabricIndex fabric, size_t index)
```
Unused, please remove.

bzbarsky-apple commented on src/messaging/ExchangeContext.h:
```diff
@@ -163,7 +163,7 @@ class DLL_EXPORT ExchangeContext : public ReliableMessageContext,
     }
 
-    bool HasSessionHandle() const { return mSession; }
+    bool HasSessionHandle() const { return (bool)mSession; }
```
Why this cast?  If a cast _is_ needed, it should be a C++-style cast.  But it should not be needed.

tcarmelveilleux replied:
That's my bad. mSession has an `operator_bool` which looks really weird when mSession is not a bool but you return a bool. I would rather that there be a method on the object, but we can just remove `(bool)` here for now.

tleacmcsa replied:
done

bzbarsky-apple commented on src/transport/SecureSession.h:
```diff
@@ -340,6 +348,8 @@ class SecureSession : public Session, public ReferenceCounted<SecureSession, Sec
     CryptoContext mCryptoContext;
     SessionMessageCounter mSessionMessageCounter;
+
+    bool mIsCaseCommissioningSession = false;
```
This should probably be placed after mSecureSessionType to pack better.

tleacmcsa replied:
done

bzbarsky-apple commented on src/transport/Session.h:
```diff
@@ -244,6 +244,12 @@ class Session
+    // Returns true if `subjectDescriptor.IsCommissioning` (based on Core Specification
+    // pseudocode in ACL Architecture chapter) should be true when computing a
+    // subject descriptor for that session.
+    virtual bool IsCommissioningSession() const { return false; }
```
Should document that this is only valid to call during synchronous processing of a message received on the session.

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessControl.cpp:
```diff
@@ -497,6 +512,45 @@ CHIP_ERROR AccessControl::Check(const SubjectDescriptor & subjectDescriptor, con
+    {
+        ChipLogProgress(DataManagement, "AccessControl: %s",
+#if 0
+                        //TODO: new error code coming with issue #35177
```
```suggestion
                        // TODO(#35177): new error code coming when access check plumbing are fixed in callers
```

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessRestrictionProvider.h:
```diff
@@ -0,0 +1,283 @@
+            virtual ~AccessRestrictionExceptionChecker() = default;
+
+            /**
+             * Check if any restrictions are allowed to be applied to the given request.
+             *
+             * @retval true if restrictions may NOT be applied
+             */
```
```suggestion
             * Check if any restrictions are disallowed to be applied to the given request
             * because of constraints against their use in ARLs.
             *
             * @retval true if ARL checks should be skipped for the path, false otherwise
             */
```

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessRestrictionProvider.h:
```diff
@@ -0,0 +1,283 @@
+            StandardAccessRestrictionExceptionChecker() = default;
+            ~StandardAccessRestrictionExceptionChecker() = default;
+
+            /**
+             * Check if any restrictions are allowed to be applied to the given request.
+             *
+             * @retval true if restrictions may NOT be applied
+             */
```
Don't recopy these docs in the derived class.

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessRestrictionProvider.h:
```diff
@@ -0,0 +1,283 @@
+         *
+         * @param [in] fabricIndex  The index of the fabric in which the entry has changed.
+         * @param [in] token        The token of the review being updated (obtained from ReviewFabricRestrictionsResponse)
+         * @param [in] instruction  The instructions to be displayed to the user.
```
This is optional so should maybe be allowed null as well.

tleacmcsa replied:
done (now Optional Span)

tcarmelveilleux commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+
+CHIP_ERROR AccessRestrictionProvider::SetCommissioningEntries(const std::vector<Entry> & entries)
+{
+    // check that the input entries are valid
```
```suggestion
```

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+{
+    std::vector<Entry> updatedEntries;
+
+    // check that the input entries are valid
```
```suggestion
```

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+                                                                         const RequestPath & requestPath)
+{
+    if (requestPath.endpoint == 0 ||
+        requestPath.cluster == app::Clusters::NetworkCommissioning::Id ||
+        requestPath.cluster == app::Clusters::Descriptor::Id)
```
Correct: let's check

- endpoint != 0
- cluster is one of the mandatory clusters in NIM

Maybe renaming to `AreRestrictionsAllowed` and reversing conditions at callsites would make this simpler

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+CHIP_ERROR AccessRestrictionProvider::CheckForCommissioning(const SubjectDescriptor & subjectDescriptor,
+                                                            const RequestPath & requestPath)
+{
+    if (mExceptionChecker.AreRestrictionsDisallowed(subjectDescriptor, requestPath))
```
This MUST be moved in `Check()` and always apply, not just in commissioning. See latest pseudo code in spec PR.

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+    // event id). All other requests require an entity id
+    if (!requestPath.entityId.has_value())
+    {
+        if (requestPath.requestType == RequestType::kEventReadOrSubscribeRequest)
```
RequestType has its elements renamed in latest spec PR. Please use the final names.

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+    return DoCheck(mFabricEntries[subjectDescriptor.fabricIndex], subjectDescriptor, requestPath);
+}
+
+CHIP_ERROR AccessRestrictionProvider::DoCheck(const std::vector<Entry> entries, const SubjectDescriptor & subjectDescriptor,
```
```suggestion
CHIP_ERROR AccessRestrictionProvider::DoCheck(const std::vector<Entry>& entries, const SubjectDescriptor & subjectDescriptor,
```

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+        }
+    }
+
+    for (auto & entry : mFabricEntries[subjectDescriptor.fabricIndex])
```
```suggestion
    for (auto & entry : entries)
```

The fact that this did not fail unit tests means that the commissioning ARL is NOT tested. Please fix

tleacmcsa replied:
fixed.  I will spend time on better tests next.

tcarmelveilleux commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+
+        for (auto & restriction : entry.restrictions)
+        {
+            // a missing id is a wildcard
```
Please ensure proper capitalization of comments throughout.

tleacmcsa replied:
Got it.

tcarmelveilleux commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+                    requestPath.requestType == RequestType::kAttributeWriteRequest)
+                {
+#if 0
+                    return CHIP_ERROR_ACCESS_RESTRICTED_BY_ARL; //TODO: new error code coming with issue #35177
```
```suggestion
                    // TODO(#35177): use new ARL error code when access checks are fixed
                    return CHIP_ERROR_ACCESS_RESTRICTED_BY_ARL;
```

tcarmelveilleux replied:
Apply the above reformatting everywhere

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+    {
+        if (!IsEntryValid(entry))
+        {
+            ChipLogError(DataManagement, "AccessRestrictionProvider: invalid entry");
```
Entry valid duplicates checks from AreRestrictionsDisallowed. Please remove the cluster checks from IsEntryValid.

tcarmelveilleux commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+
+bool AccessRestrictionProvider::IsEntryValid(const Entry & entry) const
+{
+    return entry.endpointNumber != 0 && entry.clusterId != app::Clusters::NetworkCommissioning::Id &&
```
+1

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessRestrictionProvider.h:
```diff
@@ -0,0 +1,283 @@
+     */
+    CHIP_ERROR RequestFabricRestrictionReview(FabricIndex fabricIndex, const std::vector<Entry> & arl, uint64_t & token)
+    {
+        token = ++mNextToken;
```
nit: this should be `mNextToken++` otherwise `mNextToken` is never the next token, but actually "the value before the next token"

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/AccessRestrictionProvider.h:
```diff
@@ -0,0 +1,283 @@
+    /**
+     * Perform the access restriction check using the given entries.
+     */
+    CHIP_ERROR DoCheck(const std::vector<Entry> entries, const SubjectDescriptor & subjectDescriptor,
```
```suggestion
    CHIP_ERROR DoCheck(const std::vector<Entry> & entries, const SubjectDescriptor & subjectDescriptor,
```

tleacmcsa replied:
done

tcarmelveilleux commented on src/app/clusters/access-control-server/ArlEncoder.cpp:
```diff
@@ -0,0 +1,152 @@
+    {
+        for (size_t i = 0; i < count; i++)
+        {
+            auto restriction = source[i];
```
+1 this avoids costly copies in loops

tleacmcsa replied:
done

tcarmelveilleux commented on src/app/clusters/access-control-server/ArlEncoder.h:
```diff
@@ -0,0 +1,118 @@
+         */
+        CHIP_ERROR EncodeForRead(TLV::TLVWriter & writer, TLV::Tag tag, FabricIndex fabric) const;
+
+        /**
+         * Encode the constructor-provided entry into the TLV writer.
+         */
+        CHIP_ERROR EncodeForWrite(TLV::TLVWriter & writer, TLV::Tag tag) const;
```
remove

tleacmcsa replied:
done

tcarmelveilleux commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -41,14 +47,28 @@ using Entry          = AccessControl::Entry;
+using ArlChangedEvent = Clusters::AccessControl::Events::AccessRestrictionEntryChanged::Type;
+using ArlReviewEvent  = Clusters::AccessControl::Events::FabricRestrictionReviewUpdate::Type;
+
+constexpr uint16_t kMaxInstructionStringLength = 512;
+
+constexpr uint16_t kMaxRedirectUrlStringLength = 256;
```
constexpr size_t

tleacmcsa replied:
These were removed for now (not yet used).  Part of https://github.com/project-chip/connectedhomeip/issues/35165

tcarmelveilleux commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -64,8 +84,17 @@ class AccessControlAttribute : public AttributeAccessInterface, public EntryList
+                        const AccessControl::Entry * entry, AccessControl::EntryListener::ChangeType changeType) override;
+
+#if CHIP_CONFIG_USE_ACCESS_RESTRICTIONS
+    void CommissioningRestrictionListChanged() override;
+
+    void RestrictionListChanged(FabricIndex fabricIndex) override;
```
```suggestion
    void MarkCommissioningRestrictionListChanged() override;

    void MarkRestrictionListChanged(FabricIndex fabricIndex) override;
```

tleacmcsa replied:
done

tcarmelveilleux commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,99 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+    auto accessRestrictionProvider = GetAccessControl().GetAccessRestrictionProvider();
+    if (accessRestrictionProvider == nullptr)
+    {
+        return CHIP_ERROR_NOT_IMPLEMENTED;
```
Please return CHIP_NO_ERROR here, which will leave the list empty (no restrictions). It is legal to support restrictions but not have started using a provider yet.

tcarmelveilleux replied:
Same comment for `ReadArl` below

tleacmcsa replied:
I had to call EncodeList and let it check for presence of the restriction provider, otherwise I got a bad response.

tcarmelveilleux commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,99 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+            // get entries for fabric
+            std::vector<AccessRestrictionProvider::Entry> entries;
+            ReturnErrorOnFailure(accessRestrictionProvider->GetEntries(fabric, entries));
+            for (auto & entry : entries)
```
const auto &

tleacmcsa replied:
done

tcarmelveilleux commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,99 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+    return;
+
+exit:
+    ChipLogError(DataManagement, "AccessControlCluster: restriction event failed %" CHIP_ERROR_FORMAT, err.Format());
```
```suggestion
    ChipLogError(DataManagement, "AccessControlCluster: restriction event failed: %" CHIP_ERROR_FORMAT, err.Format());
```

tleacmcsa replied:
done

tcarmelveilleux commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,99 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+
+    if (instruction != nullptr)
+    {
+        event.instruction.SetNonNull(chip::CharSpan(instruction, strnlen(instruction, kMaxInstructionStringLength)));
```
Please use CharSpan as input rather than const char* and that solves this strnlen problem for you...

tleacmcsa replied:
done

tcarmelveilleux commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -428,6 +467,99 @@ void AccessControlAttribute::OnEntryChanged(const SubjectDescriptor * subjectDes
+    return;
+
+exit:
+    ChipLogError(DataManagement, "AccessControlCluster: review event failed %" CHIP_ERROR_FORMAT, err.Format());
```
```suggestion
    ChipLogError(DataManagement, "AccessControlCluster: review event failed: %" CHIP_ERROR_FORMAT, err.Format());
```

tleacmcsa replied:
done

tcarmelveilleux commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -481,4 +613,72 @@ void MatterAccessControlPluginServerInitCallback()
+    auto entryIter = commandData.arl.begin();
+    while (entryIter.Next())
+    {
+        AccessRestrictionProvider::Entry entry;
+        entry.fabricIndex    = commandObj->GetAccessingFabricIndex();
+        entry.endpointNumber = entryIter.GetValue().endpoint;
+        entry.clusterId      = entryIter.GetValue().cluster;
+
```
I don't think this is possible/feasible as these entries may be different than those currently in the ARL.

The spec must be updated so that each ARL entry has a unique handle, and the request has a list of handles that are to be reviewed. Otherwise it's extremely error prone and it's possible the entries cannot be reviewed or cryptic errors will arise.

tleacmcsa replied:
Tracked in https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/10338

tleacmcsa committed:
Updates for review comments

tleacmcsa committed:
Merge remote-tracking branch 'upstream/master' into arl-feature

tleacmcsa committed:
restyled

tleacmcsa commented on examples/platform/linux/ExampleAccessRestrictionProvider.h:
```diff
@@ -0,0 +1,55 @@
+        chip::app::Clusters::AccessControl::Events::FabricRestrictionReviewUpdate::Type event{ .token       = token,
+                                                                                               .fabricIndex = fabricIndex };
+        EventNumber eventNumber;
+        ReturnErrorOnFailure(chip::app::LogEvent(event, 0, eventNumber));
```
done

tleacmcsa commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+{
+    if (requestPath.endpoint == 0 ||
+        requestPath.cluster == app::Clusters::NetworkCommissioning::Id ||
+        requestPath.cluster == app::Clusters::Descriptor::Id)
```
done

tleacmcsa commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+CHIP_ERROR AccessRestrictionProvider::DoCheck(const std::vector<Entry> entries, const SubjectDescriptor & subjectDescriptor,
+                                              const RequestPath & requestPath)
+{
+    ChipLogProgress(DataManagement, "AccessRestrictionProvider: action %d", static_cast<int>(requestPath.requestType));
```
done

tleacmcsa commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,252 @@
+
+bool AccessRestrictionProvider::IsEntryValid(const Entry & entry) const
+{
+    return entry.endpointNumber != 0 && entry.clusterId != app::Clusters::NetworkCommissioning::Id &&
```
done

tcarmelveilleux commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -628,9 +610,10 @@ bool emberAfAccessControlClusterReviewFabricRestrictionsCallback(
+    if (commandPath.mEndpointId != kRootEndpointId)
     {
         ChipLogError(DataManagement, "AccessControlCluster: invalid endpoint in ReviewFabricRestrictions request");
+        commandObj->AddStatus(commandPath, Protocols::InteractionModel::Status::InvalidValue);
```
InvalidCommand is likely the right error.

tleacmcsa replied:
done

tleacmcsa committed:
Review updates

- fixed return type for some command failures
- enhanced unit tests

tleacmcsa committed:
Merge remote-tracking branch 'upstream/master' into arl-feature

tleacmcsa committed:
restyled

tcarmelveilleux commented on src/access/tests/TestAccessRestrictionProvider.cpp:
```diff
@@ -207,49 +212,103 @@ TEST_F(TestAccessRestriction, MetaTest)
+    EXPECT_EQ(accessRestrictionProvider.SetEntries(1, entries), CHIP_NO_ERROR);
+
+    // also test a cluster on endpoint 0 that isnt in the special allowed list
+    entries.clear();
+    entry.clusterId = kOnOffCluster;
+    entries.push_back(entry);
+    EXPECT_EQ(accessRestrictionProvider.SetEntries(1, entries), CHIP_ERROR_INVALID_ARGUMENT);
+}
```
This doesn't actually set the entry on endpoint 0. This should not fail...

tleacmcsa replied:
done

tcarmelveilleux commented on src/access/tests/TestAccessRestrictionProvider.cpp:
```diff
@@ -470,5 +529,85 @@ TEST_F(TestAccessRestriction, CombinedRestrictionTest)
+    std::vector<AccessRestrictionProvider::Entry> arlEntriesFetched;
+    EXPECT_EQ(accessRestrictionProvider.GetEntries(2, arlEntriesFetched), CHIP_NO_ERROR);
+
+    EXPECT_NE(commissioningEntriesFetched[0].fabricIndex, arlEntriesFetched[0].fabricIndex);
+    EXPECT_NE(commissioningEntriesFetched[0].endpointNumber, arlEntriesFetched[0].endpointNumber);
+    EXPECT_NE(commissioningEntriesFetched[0].clusterId, arlEntriesFetched[0].clusterId);
+    EXPECT_NE(commissioningEntriesFetched[0].restrictions[0].restrictionType, arlEntriesFetched[0].restrictions[0].restrictionType);
+    EXPECT_NE(commissioningEntriesFetched[0].restrictions[0].id, arlEntriesFetched[0].restrictions[0].id);
```
- Missing the positive check with EXPECT_EQ that the entries are those expected for the fabric requested.
- Suggest operator== so that the entirety of entries can be compared without fully expanded individuyal field checks.

tcarmelveilleux commented on src/access/tests/TestAccessRestrictionProvider.cpp:
```diff
@@ -470,5 +529,85 @@ TEST_F(TestAccessRestriction, CombinedRestrictionTest)
+    EXPECT_NE(commissioningEntriesFetched[0].restrictions[0].id, arlEntriesFetched[0].restrictions[0].id);
+}
+
+constexpr CheckData listSelectionDuringCommissioningData[] = {
```
- `kListXxx` --> `k` prefix for constants

tleacmcsa replied:
Please let me know if you feel strongly about this.  It doesnt match others like AccessControlTest.

tleacmcsa replied:
Added to https://github.com/project-chip/connectedhomeip/issues/35165

tleacmcsa committed:
Merge remote-tracking branch 'upstream/master' into arl-feature

tleacmcsa committed:
Updated ARL tests per review comments

bzbarsky-apple commented on src/access/AccessRestrictionProvider.cpp:
```diff
@@ -0,0 +1,249 @@
+bool AccessRestrictionProvider::StandardAccessRestrictionExceptionChecker::AreRestrictionsAllowed(EndpointId endpoint,
+                                                                                                  ClusterId cluster)
+{
+    if (endpoint != 0 &&
```
```suggestion
    if (endpoint != kRootEndpointId &&
```

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -481,4 +595,81 @@ void MatterAccessControlPluginServerInitCallback()
+        entry.endpointNumber = entryIter.GetValue().endpoint;
+        entry.clusterId      = entryIter.GetValue().cluster;
+
+        auto restrictionIter = entryIter.GetValue().restrictions.begin();
```
So as you said, there are two iterators here: `entryIter` and `restrictionIter`.  Both are iterators over TLV lists, and both need the GetStatus checks.  The one for `entryIter` got added.  The one for `restrictionIter` did not and needs to be added.

In you said in https://github.com/project-chip/connectedhomeip/pull/34932#discussion_r1730430270 :

> You had mentioned some additional complexity around loading more entries.

I think I was talking about the "Load More" links in the github UI, not anything about this code....

tleacmcsa committed:
work around nuttx and jsoncpp contention

tleacmcsa committed:
Merge branch 'master' into arl-feature

tleacmcsa committed:
Review comments and nuttx build failure fix attempt

bzbarsky-apple commented on src/app/clusters/access-control-server/access-control-server.cpp:
```diff
@@ -645,6 +645,13 @@ bool emberAfAccessControlClusterReviewFabricRestrictionsCallback(
+        if (restrictionIter.GetStatus() != CHIP_NO_ERROR)
+        {
+            ChipLogError(DataManagement, "AccessControlCluster: invalid ARL data");
+            commandObj->AddStatus(commandPath, Protocols::InteractionModel::Status::InvalidValue);
```
```suggestion
            commandObj->AddStatus(commandPath, Protocols::InteractionModel::Status::InvalidCommand);
```

since I think this falls under:

> incoming data cannot be mapped to the expected data type for a field

in the spec.  The other GetStatus() check should be doing the same; I had missed that, sorry.

tleacmcsa replied:
Ahh, thanks for showing me what I missed in the spec.  Fixed.

tleacmcsa committed:
review updates

andy31415 commented on src/transport/SecureSession.cpp:
```diff
@@ -183,6 +185,24 @@ Access::SubjectDescriptor SecureSession::GetSubjectDescriptor() const
+
+    // CASE session is a commissioning session if it was marked as such.
+    // The SessionManager is what keeps track.
+    if (IsCASESession() && mIsCaseCommissioningSession)
+    {
+        return true;
+    }
+
+    return false;
```
This should be:

```suggestion
    return IsCASESession() && mIsCaseCommissioningSession;
```

andy31415 commented on src/app/clusters/access-control-server/ArlEncoder.h:
```diff
@@ -0,0 +1,113 @@
+        CHIP_ERROR Stage() const;
+
+        Entry mEntry;
+        mutable StagingEntry mStagingEntry;
```
Please remove mutable usage as a followup

andy31415 commented on examples/platform/linux/ExampleAccessRestrictionProvider.h:
```diff
@@ -0,0 +1,55 @@
+        chip::app::Clusters::AccessControl::Events::FabricRestrictionReviewUpdate::Type event{ .token       = token,
+                                                                                               .fabricIndex = fabricIndex };
+        EventNumber eventNumber;
+        ReturnErrorOnFailure(chip::app::LogEvent(event, kRootEndpointId, eventNumber));
+
+        return CHIP_NO_ERROR;
```
```suggestion
        return chip::app::LogEvent(event, kRootEndpointId, eventNumber);
```

andy31415 commented on src/access/AccessControl.cpp:
```diff
@@ -497,6 +512,45 @@ CHIP_ERROR AccessControl::Check(const SubjectDescriptor & subjectDescriptor, con
+#else
+                        (result == CHIP_ERROR_ACCESS_DENIED) ? "denied (restricted)" : "denied (restriction error)");
+#endif
+        return result;
+    }
+
+    return result;
```
```suggestion
    }
    
    return result;
```

andy31415 commented on src/app/clusters/access-control-server/ArlEncoder.cpp:
```diff
@@ -0,0 +1,145 @@
+CHIP_ERROR ArlEncoder::CommissioningEncodableEntry::Encode(TLV::TLVWriter & writer, TLV::Tag tag) const
+{
+    ReturnErrorOnFailure(Stage());
+    ReturnErrorOnFailure(mStagingEntry.Encode(writer, tag));
+    return CHIP_NO_ERROR;
```
Throughout this should look like:

```suggestion
    return mStagingEntry.Encode(writer, tag);
```

andy31415 commented on src/app/clusters/access-control-server/ArlEncoder.cpp:
```diff
@@ -0,0 +1,145 @@
+    mStagingEntry.cluster  = mEntry.clusterId;
+    ReturnErrorOnFailure(StageEntryRestrictions(mEntry.restrictions, mStagingRestrictions,
+                                                sizeof(mStagingRestrictions) / sizeof(mStagingRestrictions[0])));
+    mStagingEntry.restrictions = Span<StagingRestriction>(mStagingRestrictions, mEntry.restrictions.size());
```
you use vector anyway. Are these frequently encoded? We can avoid mutable by letting the ToTLV encode things to a vector and then use a span into that vector and not require members of fixed sizes.
